doc/imageview_warmups.md

```diff
@@ -0,0 +1,329 @@
+bugs.
+
+So in the spirit of exploration, let's look at a second implementation
+that preserves history:
+
+```jl
+module TwoNodes
+
+using Reactive
+
+export Canvas, state1!, state2!
+
+type Canvas{S<:IO}
+    io::S
+    state1::Node{Int}
+    state2::Node{Int}
+
+    function Canvas(io::S, s1::Integer, s2::Integer)
+        n1, n2 = Node(Int, s1), Node(Int, s2)
+        c = new(io, n1, n2)
+        combined = merge(n1, n2)
+        throttled = throttle(1/60, combined)
+        Reactive.preserve(map(x->println(c.io, "state1: $(value(c.state1)); state2: $(value(c.state2))"), throttled))
+        c
+    end
+end
+
+Canvas(io::IO, s1, s2) = Canvas{typeof(io)}(io, s1, s2)
+
+state1!(c::Canvas, val) = push!(c.state1, val)
+state2!(c::Canvas, val) = push!(c.state2, val)
+
+end  # module
+
+using Reactive, TwoNodes
+
+tmr = Timer(_->Reactive.run_till_now(), 0.001, 0.001)
+
+c = Canvas(STDOUT, 1, 1)
+state1!(c, 5)
+sleep(0.1)
+
+state1!(c, 7)
+state2!(c, -3)
+```
+
+Note that the implementation of the `state!` functions was simpler
+here.  With `throttle` we again get the same output:
+
+```jl
+julia> include("twonodes.jl")
+state1: 1; state2: 1
+state1: 5; state2: 1
+
+julia> state1: 7; state2: -3
+julia>
+```
+
+but this time, without `throttle` we get output that respects the history:
+
+```jl
+julia> include("twonodes.jl")
+state1: 1; state2: 1
+state1: 5; state2: 1
+state1: 7; state2: 1
+
+julia> state1: 7; state2: -3
+julia>
+```
+
+## Example 2: sharing state
+
+Suppose we have two `Canvas`es that we want to couple together: for
+example, you might want to show two views of the same image, one in
+"raw" form and the other "annotated" by some kind of image processing
+algorithm.  If you zoom in on one canvas, you might like to
+automatically zoom in on the same region in the other canvas.
+
+```jl
+module TwoSharedNodes
+
+using Reactive
+
+export Canvas, state1!, state2!, relink!
+
+type Canvas{S<:IO}
+    io::S
+    name::ASCIIString
+    state1::Node{Int}
+    state2::Node{Int}
+    update::Node{Void}
+
+    Canvas(io::S, name, s1, s2) = Canvas(io, name, node(Int, s1), node(Int, s2))
+    function Canvas(io::S, name, n1::Node{Int}, n2::Node{Int})
+        c = new(io, name, n1, n2)
+        relink!(c)
+    end
+end
+
+node{T}(::Type{T}, val) = Node(T, val)
+node{T}(::Type{T}, n::Node{T}) = n
+
+function relink!(c::Canvas)
+    isdefined(c, :update) && close(c.update)
+    combined = merge(c.state1, c.state2)
+    throttled = throttle(1/60, combined)
+    c.update = map(x->println(c.io, "$(c.name): state1=$(value(c.state1)), state2=$(value(c.state2))"), throttled)
+    c
+end
+
+Canvas(io::IO, name, s1, s2) = Canvas{typeof(io)}(io, name, s1, s2)
+
+state1!(c::Canvas, val) = push!(c.state1, val)
+state2!(c::Canvas, val) = push!(c.state2, val)
+
+end  # module
+
+using Reactive, TwoSharedNodes
+
+tmr = Timer(_->Reactive.run_till_now(), 0.001, 0.001)
+
+n2 = Node(22)
+c1 = Canvas(STDOUT, "canvas1", 1, n2)
+c2 = Canvas(STDOUT, "canvas2", 2, n2)
```

Comment:
> I guess, it would be nice to have a way to link and unlink nodes in Reactive? so that when linked, updating one of them also updates the other...

If it's easy to do, it seems like it could be nice. I do this in [GtkUtilities](https://github.com/timholy/GtkUtilities.jl#managing-state). For example, when I have a movie I'm looking at with 4000 frames in it, the slider is nice for uses but the entry box is best if you really need to go to frame 1374 and see what weird thing happened there...

GtkUtilities' API overall is not as flexible as Reactive. However, it is good for these kinds of "shared state" problems.

> I'm trying to understand this better, would it be fair to say that the code after this line is a simulation of user interacting with the Canvas UI?

Mostly "yes." One of the things I like about Reactive is that it seems to simplify having two points of control; this could be either "UI" or programmatic. But the main intent is to simulate user-driven updates. So each setting of `state!` might correspond to a click or drag in the UI; the "disconnect" could be triggered, for example, by right-clicking and picking an item on a popup menu.